flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/forwardregion/ForwardRegionComputeUtil.java

```diff
@@ -0,0 +1,89 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.adaptivebatch.forwardregion;
+
+import org.apache.flink.runtime.executiongraph.RegionComputeUtil;
+import org.apache.flink.runtime.jobgraph.JobEdge;
+import org.apache.flink.runtime.jobgraph.JobVertex;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/** Common utils for computing forward regions. */
+public class ForwardRegionComputeUtil {
+
+    public static Map<JobVertexID, ForwardRegion> computeForwardRegions(
+            final Iterable<JobVertex> topologicallySortedVertices) {
+
+        final Map<JobVertex, Set<JobVertex>> vertexToRegion = new IdentityHashMap<>();
+
+        // iterate all the vertices which are topologically sorted
+        for (JobVertex vertex : topologicallySortedVertices) {
+            Set<JobVertex> currentRegion = new HashSet<>();
+            currentRegion.add(vertex);
+            vertexToRegion.put(vertex, currentRegion);
+
+            for (JobEdge input : getForwardInputs(vertex)) {
+                final JobVertex producerVertex = input.getSource().getProducer();
+                final Set<JobVertex> producerRegion = vertexToRegion.get(producerVertex);
+
+                if (producerRegion == null) {
+                    throw new IllegalStateException(
+                            "Producer task "
+                                    + producerVertex.getID()
+                                    + " forward region is null"
+                                    + " while calculating forward region for the consumer task "
+                                    + vertex.getID()
+                                    + ". This should be a forward region building bug.");
+                }
+
+                if (currentRegion != producerRegion) {
+                    currentRegion =
+                            RegionComputeUtil.mergeRegions(
+                                    currentRegion, producerRegion, vertexToRegion);
+                }
+            }
+        }
+
+        Set<ForwardRegion> forwardRegions =
+                RegionComputeUtil.uniqueRegions(vertexToRegion).stream()
+                        .map(ForwardRegion::new)
+                        .collect(Collectors.toSet());
```

Comment:
I would suggest to exclude regions which only has one vertex, which means it is actually in no forward region. This allows use to avoid keeping unnecessary information.